pkg/api/v1alpha1/nodeupgrade_types.go

```diff
@@ -1,28 +1,62 @@
 package v1alpha1
 
 import (
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 )
 
-// NodeUpgradeKind stores the Kind for NodeUpgrade.
-const NodeUpgradeKind = "NodeUpgrade"
+const (
+	// NodeUpgradeKind stores the Kind for NodeUpgrade.
+	NodeUpgradeKind = "NodeUpgrade"
+
+	// UpgraderPodCreated reports whether the upgrader pod has been created for the node upgrade.
+	UpgraderPodCreated ConditionType = "UpgraderPodCreated"
```

Comment:
I don't know if it's wise making the phases of an upgrade tied in this way to the API. Your phases might change more frequently than you want your API to change. Also, what if some phases become optional? How are client going to know that?

Reply:
>I don't know if it's wise making the phases of an upgrade tied in this way to the API. Your phases might change more frequently than you want your API to change.

Can you elaborate a little on this? This is just adding container statuses on the node upgrade object.
The controller will check the container statuses at the end of every reconcile loop and update the node upgrade status based on that. And once all the containers have finished, it will mark the upgrade as completed and end the reconcile loop.
I extrapolated  this logic from the EKS-A controller which does something similar where it monitors the KCP/MD/CAPI cluster conditions and uses it to update EKS-A Cluster conditions. And when all conditions are met, the reconcile is marked complete.

>Also, what if some phases become optional? How are client going to know that?

This depends on how we decide to implement it. We were thinking we won't make the phases optional from NodeUpgradeController's perspective. Rather the pod will check if the step it's performing needs to be done or not. If it doesn't need to be performed, the upgrader will just not do it and return a success. 
For example, if the containerd version on the node is already the latest one, then the upgrader can just skip copying over the containerd binary and return without errors.